Sum type cannot be passed as mutable parameter without reference

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version:** 
**OS:** 

**What did you do?**

```v
type Sum = Struct1 | int

struct Struct1 {
mut:
	value int
}

fn update_sum(mut sum Sum) {
	match mut sum {
		Struct1 {
			sum.value = 42
		}
		else {}
	}
}

fn main() {
	mut s := Sum(Struct1{
		value: 6
	})

	update_sum(mut s)
	if mut s is Struct1 {
		println(s.value)
	}
}
```


**What did you expect to see?**
Working program, printing `42`
 
**What did you see instead?**

```
==================
/tmp/v_1000/error4.3847824426243062934.tmp.c:11064: error: cannot convert 'struct main__Sum' to 'struct main__Sum *'
...
==================
(Use `v -cg` to print the entire error message)
```

One can work around this by adding an explicit reference to `update(mut &s)`, but V should do this automatically in the same way as with structs.

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -1602,6 +1602,7 @@ fn (mut g Gen) ref_or_deref_arg(arg ast.CallArg, expected_type ast.Type, lang as
 		g.write('&/*mut*/')
 	} else if arg_is_ptr && !expr_is_ptr {
 		if arg.is_mut {
+			arg_sym := g.table.sym(arg_typ)
 			if exp_sym.kind == .array {
 				if (arg.expr is ast.Ident && (arg.expr as ast.Ident).kind == .variable)
 					|| arg.expr is ast.SelectorExpr {
@@ -1614,6 +1615,11 @@ fn (mut g Gen) ref_or_deref_arg(arg ast.CallArg, expected_type ast.Type, lang as
 					g.expr(arg.expr)
 					g.write('}[0]')
 				}
+				return
+			} else if arg_sym.kind == .sum_type && exp_sym.kind == .sum_type
+				&& arg.expr is ast.Ident {
+				g.write('&')
+				g.expr(arg.expr)
 				return
 			}
 		}

--- a/vlib/v/tests/fn_call_mut_sumtype_args_test.v
+++ b/vlib/v/tests/fn_call_mut_sumtype_args_test.v
@@ -1,0 +1,30 @@
+type Sum = Struct1 | int
+
+struct Struct1 {
+mut:
+	value int
+}
+
+fn update_sum(mut sum Sum) {
+	match mut sum {
+		Struct1 {
+			sum.value = 42
+		}
+		else {}
+	}
+}
+
+fn test_fn_call_mut_sumtype_args() {
+	mut s := Sum(Struct1{
+		value: 6
+	})
+
+	update_sum(mut s)
+
+	if mut s is Struct1 {
+		println(s.value)
+		assert s.value == 42
+	} else {
+		assert false
+	}
+}

